docs/dev/virtual_machine.rst

```diff
@@ -0,0 +1,318 @@
+programs to this we translate a small subset of them to the old graph format, and execute
+them on the runtime.
+This provides a solid execution experience but only for a very limited subset of Relay programs.
+
+An alternative but not-standard approach is Relay's ahead-of-time compiler
+which transforms a Relay program into a shared library containing an ahead
+of time implementation. The ahead of time compiler provides compelling performance
+but is difficult to extend, and instrument, requiring modifications to the
+code generation and optimizations.
+
+The Relay virtual machine is intended to be a framework that balances these competing
+approaches providing a dynamic execution environment which can be extended, instrumented,
+and integrated with other approaches like ahead of time compilation via a flexible extension
+mechanism.
+
+The virtual machine is designed to strike a balance between performance and flexibility
+when deploying and executing Relay programs, without giving up the benefits of TVM.
+
+Virtual machine (VM) design is a well studied area in programming languages and systems,
+and there have been various virtual machine designs for both full fledged,
+and embedded programing languages.
+Previous language VM designs have been heavily tailored to the execution profile of traditional programs.
+Traditional programs manipulate small scalar values and consist of a large number of low level instructions.
+The sheer quantity of instructions to compute requires instruction execution and dispatch to be extremely efficient.
+In the context of machine learning we manipulate primarily tensor values, using a (relatively)
+low number of high level instructions. ML program's cost centers are expensive operator invocations
+such as GEMM or convolution, over a large input. Due to the execution profile exhibited by ML programs
+micro-optimizations present in scalar-VMs are dramatically less important.
+A model’s runtime will  be dominated by executing expensive operators on large inputs.
+
+TVM has provided a strong support for vision models,
+but we want to grow to support a wider variety of models.
+The graph runtime is able to utilize the fully static nature of the input graphs to perform
+aggressive optimization such as fully static allocation, and optimal memory reuse.
+When we introduce models which make use of control-flow, recursion, dynamic shapes, and dynamic
+allocation, we must change how execution works.
+
+The rest of this document provides a high level overview of the Relay
+virtual machine design, and its instruction set.
+
+Design
+------
+
+The VM's design is focused on simplicity without sacrificing performance.
+In order to accomplish this we have ignored traditional wisdom in scalar
+VM design, and focused on designing a tensor VM.
+
+In the tensor VM setting, we optimize for cheap “allocation” of objects (by trying to avoid real allocation,
+reuse of static fragments, and the ability to do dynamic (i.e jagged tensors)).
```

Comment:
The way this sentence is written is confusing.  It seems like you're saying you _avoid_ reuse of static fragments and the ability to handle jagged tensors.